### PR TITLE
feat:  Add configuration allowItemToFail to poll mode, and add a gate…

### PR DESCRIPTION
**Pull request by dkichler**
… around each access_token/item update to allow it to fail without failing the entire poll sync

Maintains backward compatibility with existing behaviour with default of `false`.  

Addressed #91

### CODE REVIEW

**Comment by dvankley** [comment]
I'm still trying to get my oauth setup with Plaid sorted out so I can fix my personal build and have something to regression test on; will take a look at this (and the rest of the backlog of issues) once that's done.

**Comment by dkichler** [comment]
Did a round of regression testing on this myself only to realize how naive I was wrt to coroutines (this was my first encounter with them, havin spend almost no time in Kotlin).  Fixed it up and verified it works as intended in my local instance.

File: src/main/kotlin/net/djvk/fireflyPlaidConnector2/sync/PolledSyncRunner.kt [hunk 1]
```diff
@@ -40,6 +41,8 @@ class PolledSyncRunner(
     private val cursorFileDirectoryPath: String,
     @Value("\${fireflyPlaidConnector2.plaid.batchSize}")
     private val plaidBatchSize: Int,
+    @Value("\${fireflyPlaidConnector2.polled.allowItemToFail}")
```
[the remark this answers is not in the document]
Comment (dvankley) [reply src/main/kotlin/net/djvk/fireflyPlaidConnector2/sync/PolledSyncRunner.kt:44]: Also, do you want this to be per-account or global? If per-account, you'd want to change `net.djvk.fireflyPlaidConnector2.config.AccountConfig` instead.

File: src/main/kotlin/net/djvk/fireflyPlaidConnector2/sync/PolledSyncRunner.kt [hunk 2]
```diff
@@ -333,4 +340,16 @@ class PolledSyncRunner(
...
+        fun emptyPlaidResponse(): TransactionsSyncResponse =
+            TransactionsSyncResponse(
+                Collections.emptyList(),
+                Collections.emptyList(),
+                Collections.emptyList(),
+                "",
+                false,
+                ""
+            )
+    }
```
Comment (dvankley) [inline src/main/kotlin/net/djvk/fireflyPlaidConnector2/sync/PolledSyncRunner.kt:354]: I'm not crazy about creating this dummy object just to allow a different return mode from that function. How about using `null` instead?

Comment (dvankley) [reply src/main/kotlin/net/djvk/fireflyPlaidConnector2/sync/PolledSyncRunner.kt:354]: I expanded on this idea a bit in https://github.com/dkichler/firefly-plaid-connector-2/pull/1

**Review by dvankley — CHANGES_REQUESTED** [review]

File: src/main/kotlin/net/djvk/fireflyPlaidConnector2/sync/PolledSyncRunner.kt (hunk 2 above)
Comment (dkichler) [reply src/main/kotlin/net/djvk/fireflyPlaidConnector2/sync/PolledSyncRunner.kt:354]: > I'm not crazy about creating this dummy object just to allow a different return mode from that function. How about using null instead?

Coming from Scala, I tend to naturally shy away from `null` - I can give it a shot but I think the overall changes wil

My goal here was to achieve the behaviour with as few changes as possible  - there are no tests covering this piece of code and so testing was limited to adhoc testing against an actual instance.

Admittedly, this approach was my second preference to throwing the exception from `executeTransactionSyncRequest` and catching it upstream, but of course that is complicated by the coroutine boundary.  Handling internal to the method and using an "empty" result was an easy way to minimize the changes. 

> I expanded on this idea a bit in https://github.com/dkichler/firefly-plaid-connector-2/pull/1

It look's like this patch also includes a some additiona refactor for behaviour over and above the intent of this PR.  

I would propose that if you're opposed to the "empty" response, I can refactor this piece to use null (I might suggest Option, but I see Kotlin doesn't have it natively), and you can post the remainder of your suggestion as a standalone PR to adjust cursor catchup behaviour.

**Review by dkichler — COMMENTED** [review]
Responded to review comments, and have a small change to post as a reduced scope of your suggestion patch (only using `null` rather than an empty response).

File: src/main/kotlin/net/djvk/fireflyPlaidConnector2/sync/PolledSyncRunner.kt (hunk 2 above)
Comment (dkichler) [reply src/main/kotlin/net/djvk/fireflyPlaidConnector2/sync/PolledSyncRunner.kt]: Swapping in `null` for the empty response wasn't too bad - changes are still pretty minimal.

File: src/main/kotlin/net/djvk/fireflyPlaidConnector2/sync/PolledSyncRunner.kt [hunk 3]
```diff
@@ -157,22 +162,28 @@ class PolledSyncRunner(
                                 cursorMap[accessToken],
                                 plaidBatchSize
                             )
```
Comment (dvankley) [inline src/main/kotlin/net/djvk/fireflyPlaidConnector2/sync/PolledSyncRunner.kt:164]: I did get a bit carried away with my suggested changes, but I do think my suggestion here has merit.

Changing line 140 to 
```
accessTokenLoop@ for ((accessToken, accountIds) in accountAccessTokenSequence) {
```

and here to 
```
) ?: continue@accessTokenLoop
```

will be cleaner than adding 10 safe call operators.

File: src/main/resources/application.yml [hunk 4]
```diff
@@ -33,6 +33,9 @@ fireflyPlaidConnector2:
     # This path needs to be writeable by the application's user and needs to be persistent. If you're using Docker,
     #   you may want to use a volume or bind mount for this so that cursor state is persisted between runs.
     cursorFileDirectoryPath: persistence/
+    # Configuration element specifying whether the remainder of the poll sync session should continue if one item
+    # has failed to fetch from PLaid for any reason.
```
Comment (dvankley) [inline src/main/resources/application.yml:37]: Typo: `PLaid`

File: src/main/kotlin/net/djvk/fireflyPlaidConnector2/sync/PolledSyncRunner.kt [hunk 5]
```diff
@@ -81,12 +84,14 @@ class PolledSyncRunner(
                             executeTransactionSyncRequest(accessToken, cursorMap[accessToken], plaidBatchSize)
```
Comment (dvankley) [inline src/main/kotlin/net/djvk/fireflyPlaidConnector2/sync/PolledSyncRunner.kt:84]: Same comment as line 164. There are less null safe calls in this case, but being consistent is preferable.

Comment (dvankley) [reply src/main/kotlin/net/djvk/fireflyPlaidConnector2/sync/PolledSyncRunner.kt:84]: You could also just throw an exception here if you get null for initial cursor sync, but I'm fine either way with that.

**Review by dvankley — CHANGES_REQUESTED** [review]

File: src/main/kotlin/net/djvk/fireflyPlaidConnector2/sync/PolledSyncRunner.kt (hunk 3 above)
Comment (dkichler) [reply src/main/kotlin/net/djvk/fireflyPlaidConnector2/sync/PolledSyncRunner.kt:164]: Aha, indeed, this is where I started with a `?let { it } ?: run { continue }` but of course continue cannot be used in a lamdba.  This is exactly the syntax I needed - good suggestion.

Comment (dkichler) [reply src/main/kotlin/net/djvk/fireflyPlaidConnector2/sync/PolledSyncRunner.kt]: Took this suggestion in both spots.

**Comment by dkichler** [comment]
Ok, all feedback addressed and the patch is nice and slim 😎

**Comment by dvankley** [comment]
Looks good. Thanks for bearing with me on the feedback cycle.
I'll merge it and cut a new release when I get a chance.

**Comment by dkichler** [comment]
No problem, it was good feedback!

**Pull request merged**